machine/htif.c

```diff
@@ -85,3 +95,44 @@ void htif_poweroff()
     tohost = 1;
   }
 }
+
+struct htif_scan
+{
+  int compat;
+};
+
+static void htif_open(const struct fdt_scan_node *node, void *extra)
+{
+  struct htif_scan *scan = (struct htif_scan *)extra;
+  memset(scan, 0, sizeof(*scan));
+}
+
+static void htif_prop(const struct fdt_scan_prop *prop, void *extra)
+{
+  struct htif_scan *scan = (struct htif_scan *)extra;
+  if (!strcmp(prop->name, "compatible") && !strcmp((const char*)prop->value, "ucb,htif0")) {
+    scan->compat = 1;
+  }
+}
+
+static void htif_done(const struct fdt_scan_node *node, void *extra)
+{
+  struct htif_scan *scan = (struct htif_scan *)extra;
+  if (!scan->compat) return;
+
+  htif = 1;
+}
+
+void query_htif(uintptr_t fdt)
+{
+  struct fdt_cb cb;
+  struct htif_scan scan;
+
+  memset(&cb, 0, sizeof(cb));
+  cb.open = htif_open;
+  cb.prop = htif_prop;
+  cb.done = htif_done;
+  cb.extra = &scan;
+
+  fdt_scan(fdt, &cb);
+}
```

Comment:
It makes me a bit sad how much code it takes to extract a property from DTS. Without using dynamic memory, tho, I'm not sure how to improve things.